Pure ESM

PowerKiKi opened the pull request:
I am trying to remove all references to non-ESM.

For that I used [dpdm](https://www.npmjs.com/package/dpdm), and ran something like:

```sh
dpdm src/main.ts | grep -F -C 10 'node_modules/highcharts/highcharts.js'
```

However this should not be merged as-is, because there are two issues left to solve:

First, `import('highcharts-custom-events')` now fails to compile. If I understand correctly because that plugin is not ESM. If that's the case, maybe we could decide that non-ESM, non Highcharts v12 compatible plugins are not supported. And find another plugin to use in our demo. But I might be missing something too... @karolkolodziej, any opinion on that ?

Second, maybe even more annoying is that we have a runtime error when importing some modules, such as map and 3d, as demonstrated in https://github.com/highcharts/highcharts/issues/21840 :-/

Any inputs are welcome...

Commit:
convert highcharts-chart.component.ts to standalone and deprecate module

Commit:
correct test file

Commit:
mv to ng17

Commit:
release 5.0.0-alpha.1

Commit:
use the new control flow

Commit:
remove polyfills.ts

Commit:
use standalone component instead of module

Commit:
update README.md

Commit:
update to node 18

Commit:
lint file

Commit:
update to es2020

Commit:
fix tslint issue

Commit:
fix tslint issue

Commit:
fix tslint issue

Commit:
use new output function instead of decorator

Commit:
use model signal to handle input/output

Commit:
use destroyRef instead of OnDestroy hook

Commit:
convert to Signal Input

Commit:
add new packages

Commit:
remove unused packages

Commit:
mv to ng 18

Commit:
mark not standalone as false

Commit:
mv to ng 19

Commit:
refactor highchart component

Commit:
generate jsdoc

Commit:
remove old packages

Commit:
update map collection

Commit:
update karma

Commit:
correct imports

Commit:
use esbuild to build the demo app

Commit:
convert demo project to standalone

Commit:
use onPush change detection

Commit:
ref local folder for test

Commit:
release v5.0.0

Commit:
move logic to directive, this give more flexibility to dev

Commit:
add doc for directive

Commit:
create css class

Commit:
create css class

Commit:
split command

Commit:
introduce injection token and provider to supply with module factory

Commit:
modify directive to support new changes

Commit:
adapt demo component

Commit:
a better approach to serve global setting and single providers

Commit:
correct demo app

Commit:
use new syntax approach to inject tokens

Commit:
lint files

Commit:
correct test component

Commit:
correct test component

Commit:
allow to provide highchart at the child level

Commit:
correct demo app

Commit:
use better approach

Commit:
finish upgrade

Commit:
correct configuration

Commit:
add test for service

Commit:
add test for directive

Commit:
revert back version so it will be changed when packing

Commit:
add possible chart to input "constructorType"

Commit:
update should be optional

Commit:
rename variable

Commit:
update jsdoc for oneToOne and update input

Commit:
patch packages

Commit:
patch packages

Commit:
move default app to new folder

Commit:
default command to my-app

Commit:
handle SSR case for directive

Commit:
add new ssr-app for showcase

Commit:
allowedCommonJsDependencies

Commit:
handle map case differently

Commit:
correct title

Commit:
increase budget

Commit:
we can reuse app for both ssr and no-ssr

Commit:
reset config

Commit:
reset config

Commit:
use signals

Commit:
correct type

Commit:
rename vars and add comments

Commit:
rename readme v4

Commit:
update readme.md

Commit:
update readme.md

Commit:
add support to Highcharts 12

Commit:
update showcase

Commit:
correct types

Commit:
correct tests

Commit:
update readme

Commit:
update readme

Commit:
correct test

Commit:
lazy load main module

Commit:
address feedbacks

Commit:
update README.md

Commit:
correct test

Commit:
switch to ESM

Commit:
switch to ESM

Commit:
add new commands

Commit:
update readme

PowerKiKi committed:
Pure ESM

This requires Highcharts at least 12.2.0 for proper support of ESM,
specifically internal dependencies across highcharts modules.

medbenmakhlouf commented on highcharts-angular/package.json:
```diff
@@ -21,7 +21,7 @@
     "@angular/common": ">=19.0.0",
     "@angular/core": ">=19.0.0",
-    "highcharts": ">=9.0.0"
+    "highcharts": ">=12.0.0"
```
@PowerKiKi , I have adapt the support for version 12, I think we can keep old version as well. unless you know a breaking change

PowerKiKi replied:
Since v12 [modules no longer require initialization](https://github.com/highcharts/highcharts/blob/master/changelog/highcharts/12.0.0.md#upgrade-notes). That is a breaking change for us. Supporting only v12 allow us to simplify our code. We no longer need to identify the version and conditionally call `module.default(Highcharts)`.

And since it's been historically relatively painless to upgrade Highcharts, I don't see why users wouldn't upgrade at the same time as they upgrade highcharts-angular. It is a rather low-effort for them, and avoid us supporting old things.

I am not able to find an official old version support policy. So in the absence of such, I can only assume that only the latest version of Highchart is supported. That would mean v11 is EOL.

@karolkolodziej, are you able to share something regarding which versions of Highcharts should be supported ?

medbenmakhlouf commented on highcharts-angular/src/lib/highcharts-chart.service.ts:
```diff
@@ -14,13 +14,9 @@ export class HighchartsChartService {
-    const results: ModuleFactory[] = await Promise.all(
-      partialConfig?.modules ? [...this.globalModules(), ...partialConfig.modules()] : this.globalModules()
-    );
-    if (version < 120000) {
```
@PowerKiKi , i think we can keep support for version 11, I dont see a big breaking change in highcharts 12.0.0 that will let us force dev using this package to upgrade version v5 .

karolkolodziej committed:
Update how to import modules in v12

karolkolodziej committed:
Update README.md

Co-authored-by: Dawid Draguła <46498046+ddragula@users.noreply.github.com>

karolkolodziej committed:
Merge pull request #396 from highcharts/readme/v12

readme/v12

PowerKiKi committed:
Review README

PowerKiKi committed:
Consistent casing of Highcharts

PowerKiKi committed:
BREAKING Dropped `callbackFunction`

Because the Angular way is to use outputs, via `chartInstance`

PowerKiKi committed:
input/output/model, and all `inject()` are strictly readonly

PowerKiKi committed:
Import type only

PowerKiKi committed:
Drop README v4

Because it will always be available on https://github.com/highcharts/highcharts-angular/blob/v4.0.1/README.md

PowerKiKi committed:
Drop unused imports

PowerKiKi committed:
Don't duplicate type declaration

PowerKiKi committed:
Tests don't expect Highcharts 11 anymore

PowerKiKi committed:
Replace tslint with eslint

Using most recommended rules

PowerKiKi committed:
`(chartInstance)` emits as soon as possible

This restores the demo page where the last chart was not drawn

PowerKiKi committed:
CI

PowerKiKi committed:
SSR can be built again

PowerKiKi committed:
fixup README

PowerKiKi committed:
Mention `chartInstance` is also emitted after destruction

PowerKiKi committed:
Link to old README

PowerKiKi committed:
Correct comment

PowerKiKi committed:
`chartInstance` does not emit anything during destruction

Because Angular signals prevent us from doing so.

PowerKiKi committed:
Fix SSR demo

PowerKiKi committed:
Do not emit `null` upon creation

This rework how things flow, so that the chart is created as soon as we
can, and is then kept up to date separately. This allows avoiding
relaying on "previous" value of chart, and instead using simpler
reactive concepts. In turn, we avoid emitting `null` via `chartInstance`
upon creation.

PowerKiKi committed:
Simplify typing

The indirection via `Chart` is not necessary anymore since we only
support pure ESM
